### PR TITLE
[WIP] Application list

### PR DESCRIPTION
**Pull request by pengx17**
WIP

- [X] App list page
- [X] App detail page basic info
- [x] Generic resource list api for application
- [x] App detail page components
- [x] Application name (link) in other resource detail page
- [ ] Tests

### CODE REVIEW

**Comment by pengx17** [comment]
@bryk @floreks @konryd since application is not available in the main stream k8s go-client, I need to update the dep per github.com/kubernetes-sigs/application branch. Not sure if I did it correctly but it results in a lot of dependency updates.

What I did:
- update `gopkg.toml` to include `github.com/kubernetes-sigs/application` and update some other version to match with the one listed in `github.com/kubernetes-sigs/application`
- run `dep ensure -v -add github.com/kubernetes-sigs/application/pkg`

**Comment by maciaszczykm** [comment]
@pengx17 https://github.com/kubernetes/dashboard/wiki/Dependency-management

**Comment by pengx17** [comment]
@maciaszczykm hi Marcin, I have followed the instructions and it is merged with pr #2985

**Comment by pengx17** [comment]
@konryd hi, Application list page (not including detail page) is added. If you want you could pull and build on you local machine and have a look.

qq20180510-185214 2x

For now @danielromlein has not yet involved and I did not know what would the uiux design will be like in its real state.

**Comment by pengx17** [comment]
app-detaik
Basic info has been added to detail page.

**Comment by pengx17** [comment]
qq20180518-142741 2x

Detail page is updated with components lists demo. Note, all resource lists only display simple meta columns for the moment.

File: src/app/frontend/application/detail/info.html [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  <kd-info-card-section>
+    <kd-object-meta-info-card object-meta="::$ctrl.application.objectMeta">
+    </kd-object-meta-info-card>
+    <kd-info-card-entry title="[[Selector|Label 'Selector' for the application's labels list on the application details page.]]">
+      <kd-labels labels="::$ctrl.application.selector"></kd-labels>
+    </kd-info-card-entry>
+    <kd-info-card-entry title="[[Type|Type 'Type' for the application's Type on the application details page.]]">
+      {{ ::$ctrl.application.type }}
+    </kd-info-card-entry>
+    <kd-info-card-entry title="[[Assembly Phase|Assembly Phase 'Assembly Phase' for the application's Assembly Phase on the application details page.]]">
```
Comment (konryd) [inline src/app/frontend/application/detail/info.html]: The assemblyPhase will be useful for computing the application's status, but it makes little sense to display it as a field.

File: src/app/frontend/application/detail/info.html [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                        title="[[Owners|Owners 'Owners' for the application's owners on the application details page.]]">
+      <div ng-repeat="owner in ::$ctrl.application.owners">{{ owner }}</div>
+    </kd-info-card-entry>
+    <kd-info-card-entry ng-if="::$ctrl.application.maintainers.length"
+                        title="[[Maintainers|Maintainers 'Maintainers' for the application's maintainers on the application details page.]]">
+      <div ng-repeat="maintainer in ::$ctrl.application.maintainers">{{ ::maintainer.name }} {{ ::maintainer.email }}</div>
+    </kd-info-card-entry>
+    <kd-info-card-entry ng-if="::$ctrl.application.notes"
+                        title="[[Notes|Notes 'Notes' for the application's notes on the application details page.]]">
+      {{ ::$ctrl.application.notes }}
```
Comment (konryd) [inline src/app/frontend/application/detail/info.html]: The notes will potentially contain multiple paragraphs. I think it's best to render it as a <pre>

File: src/app/frontend/application/genericresourcelist/cardlist.html [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+-->
+
+<kd-resource-card-list selectable="false"
+                       with-statuses="false"
+                       select-id="{{::$ctrl.getSelectId()}}"
+                       list="$ctrl.applicationGenericResourceList"
+                       list-resource="::$ctrl.applicationGenericResourceListResource">
+  <kd-resource-card-list-header>
+    <kd-resource-card-list-title>
+      {{ ::$ctrl.groupName }}.{{ ::$ctrl.kindName }}
```
Comment (konryd) [inline src/app/frontend/application/genericresourcelist/cardlist.html]: Let's drop the groupName from here.

File: src/app/frontend/application/detail/info.html (hunk 2 above)
Comment (konryd) [reply src/app/frontend/application/detail/info.html]: Same for description.

**Comment by danielromlein** [comment]
My main point of feedback is that our card design does not accommodate multi-line values well, and perhaps we should look to redesign to have both keys and values be L-aligned within two columns.

Current: 
image

(readability suffers)

Possible alternative:
screen shot 2018-05-29 at 3 38 29 pm

Nit: it should be "Edit *an* Application" (not "a").
image

Otherwise LGTM!

**Comment by konryd** [comment]
FYI: there is a breaking change coming from the application repo:
https://github.com/kubernetes-sigs/application/pull/45#pullrequestreview-126668119

**Comment by yinhaijiao** [comment]
@konryd @danielromlein 
I have appended it at the end of the component named 'kd-object-meta-info-card'. 
Is the layout in the picture above reasonable?

**Comment by konryd** [comment]
Looks good to me. The spacing seems a bit off, but that's most likely the result of the red frame :)

@danielromlein  ?

**Comment by yinhaijiao** [comment]
@konryd
Without red frame.

**Comment by konryd** [comment]
Thanks!

**Comment by pengx17** [comment]
> FYI: there is a breaking change coming from the application repo:
> kubernetes-sigs/application#45 (review)

@konryd I see the [pr#45](kubernetes-sigs/application#45) is still WIP. I guess there maybe a sigs/application client update soon in the future? Also, is there an estimated time when would application get merged into the main go client?

**Comment by konryd** [comment]
@ant31 FYI

**Comment by pengx17** [comment]
@danielromlein Sorry currently the changes to both your suggestions are not trivial at this stage:

> My main point of feedback is that our card design does not accommodate multi-line values well, and perhaps we should look to redesign to have both keys and values be L-aligned within two columns.

Key-value entries are row based at the moment and we cannot simply assume a minimum width of the key fields. Also the ui change would apply globally, I guess this should be done in a separate pr.

> 'Edit a application' ->  'Edit an application'
 `Edit a $resource_name` is the current prefix in the implementation (English locale). It is not trivial to dynamically convert `a` to `an` depending on the lexical context. I guess a naive change is to change this to 'Edit a(n)' to make this grammar correctly?

**Comment by danielromlein** [comment]
@pengx17 
> Sorry currently the changes to both your suggestions are not trivial at this stage

That makes sense! Not worth blocking the launch then. Sorry for the delay in response.

**Comment by danielromlein** [comment]
@pengx17 is this still being worked on, or did we decide to postpone until after the Angular migration is completed?

**Comment by pengx17** [comment]
@danielromlein I believe there are still some issues related to fixing unit tests and this is transferred to @yinhaijiao. He agreed to finish the tasks gradually, but not sure when it will be done though.

This feature should be refactored to Angular with ease since the work has been done o the backend side could be reused. I also think we might postpone this feature once migration is done, but let’s wait and see how the progress will be on Yin’s side.

**Comment by maciaszczykm** [comment]
Unfortunately, I have to close this as stale. We need to implement it using the newest Angular.

Please reopen if you are going to update it on your own.

/close

**Pull request closed without merging**